File: hooks/scripts/cmd/teammate_idle.go
// TeammateIdle: Inject available task context when teammate goes idle.
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/spf13/cobra"
	"solon-hooks/internal/config"
	"solon-hooks/internal/hookio"
)

var teammateIdleCmd = &cobra.Command{
	Use:   "teammate-idle",
	Short: "Handle TeammateIdle hook",
	RunE:  runTeammateIdle,
}

// idleTaskFile is the shape of a task file including blocking/ownership fields.
type idleTaskFile struct {
	ID        string   `json:"id"`
	Status    string   `json:"status"`
	Subject   string   `json:"subject"`
	BlockedBy []string `json:"blockedBy"`
	Owner     string   `json:"owner"`
}

type availableTaskInfo struct {
	Pending    int
	InProgress int
	Completed  int
	Total      int
	Unblocked  []struct{ ID, Subject string }
}

// getAvailableTasks reads all task files for a team and computes unblocked tasks.
func getAvailableTasks(teamName string) *availableTaskInfo {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil
	}
	taskDir := filepath.Join(home, ".claude", "tasks", teamName)
	entries, err := os.ReadDir(taskDir)
	if err != nil {
		return nil
	}

	var tasks []idleTaskFile
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		data, err := os.ReadFile(filepath.Join(taskDir, e.Name()))
		if err != nil {
			continue
		}
		var t idleTaskFile
		if err := json.Unmarshal(data, &t); err != nil || t.Status == "" {
			continue
		}
		tasks = append(tasks, t)
	}

	// Build completed ID set
	completedIDs := make(map[string]bool)
	for _, t := range tasks {
		if t.Status == "completed" {
			completedIDs[t.ID] = true
		}
	}

	info := &availableTaskInfo{}
	for _, t := range tasks {
		switch t.Status {
		case "completed":
			info.Completed++
			continue
		case "in_progress":
			info.InProgress++
			continue
		case "pending":
			info.Pending++
		default:
			continue
		}
		// Check if unblocked and unowned
		allDepsComplete := true
		for _, depID := range t.BlockedBy {
			if !completedIDs[depID] {
				allDepsComplete = false
				break
			}
		}
		if allDepsComplete && t.Owner == "" {
			info.Unblocked = append(info.Unblocked, struct{ ID, Subject string }{t.ID, t.Subject})
		}
	}
	info.Total = info.Pending + info.InProgress + info.Completed
	return info
}

func runTeammateIdle(cmd *cobra.Command, args []string) error {
	if !config.IsHookEnabled("teammate-idle-handler") {
		os.Exit(0)
	}

	defer func() {
		if r := recover(); r != nil {
			os.Exit(0)
		}
	}()

	var payload hookio.TeammateIdleInput
	if err := hookio.ReadInput(&payload); err != nil {
		os.Exit(0)
	}

	if payload.TeamName == "" {
		os.Exit(0)
	}

	info := getAvailableTasks(payload.TeamName)
	lines := []string{
		"## Teammate Idle",
		fmt.Sprintf("%s is idle.", payload.TeammateName),
	}

	if info != nil {
		remaining := info.Pending + info.InProgress
		lines = append(lines, fmt.Sprintf("Tasks: %d/%d done. %d remaining.", info.Completed, info.Total, remaining))

		if len(info.Unblocked) > 0 {
			parts := make([]string, len(info.Unblocked))
			for i, t := range info.Unblocked {
				parts[i] = fmt.Sprintf("#%s \"%s\"", t.ID, t.Subject)
			}
			lines = append(lines, "Unblocked & unassigned: "+strings.Join(parts, ", "))
			lines = append(lines, fmt.Sprintf("Consider assigning work to %s or waking them with a message.", payload.TeammateName))
		} else if remaining == 0 {
			lines = append(lines, fmt.Sprintf("No remaining tasks. Consider shutting down %s.", payload.TeammateName))
		} else {
			lines = append(lines, fmt.Sprintf("All remaining tasks are blocked or assigned. %s may be waiting for dependencies.", payload.TeammateName))
		}
	}

	hookio.WriteOutput(map[string]interface{}{
		"hookSpecificOutput": map[string]interface{}{
			"hookEventName":     "TeammateIdle",
			"additionalContext": strings.Join(lines, "\n"),
		},
	})
	return nil
}

File: crates/solon-core/src/hooks/teammate_idle.rs
use crate::hooks::{get_str, is_hook_enabled};
/// TeammateIdle hook: inject available task context when a teammate goes idle.
use anyhow::Result;

pub fn run() -> Result<()> {
    if !is_hook_enabled("teammate-idle-handler") {
        std::process::exit(0);
    }

    let input = match crate::hooks::read_hook_input() {
        Ok(v) => v,
        Err(_) => std::process::exit(0),
    };

    let team_name = get_str(&input, "team_name").to_string();
    if team_name.is_empty() {
        std::process::exit(0);
    }

    let teammate_name = get_str(&input, "teammate_name").to_string();
    let info = get_available_tasks(&team_name);

    let mut lines = vec![
        "## Teammate Idle".to_string(),
        format!("{} is idle.", teammate_name),
    ];

    if let Some(info) = &info {
        let remaining = info.pending + info.in_progress;
        lines.push(format!(
            "Tasks: {}/{} done. {} remaining.",
            info.completed, info.total, remaining
        ));

        if !info.unblocked.is_empty() {
            let parts: Vec<String> = info
                .unblocked
                .iter()
                .map(|(id, subject)| format!("#{} \"{}\"", id, subject))
                .collect();
            lines.push(format!("Unblocked & unassigned: {}", parts.join(", ")));
            lines.push(format!(
                "Consider assigning work to {} or waking them with a message.",
                teammate_name
            ));
        } else if remaining == 0 {
            lines.push(format!(
                "No remaining tasks. Consider shutting down {}.",
                teammate_name
            ));
        } else {
            lines.push(format!(
                "All remaining tasks are blocked or assigned. {} may be waiting for dependencies.",
                teammate_name
            ));
        }
    }

    crate::hooks::write_output(&serde_json::json!({
        "hookSpecificOutput": {
            "hookEventName": "TeammateIdle",
            "additionalContext": lines.join("\n"),
        }
    }));
    Ok(())
}

struct TeamInfo {
    pending: usize,
    in_progress: usize,
    completed: usize,
    total: usize,
    unblocked: Vec<(String, String)>, // (id, subject)
}

#[derive(serde::Deserialize)]
struct TaskFile {
    #[serde(default)]
    id: String,
    #[serde(default)]
    status: String,
    #[serde(default)]
    subject: String,
    #[serde(rename = "blockedBy", default)]
    blocked_by: Vec<String>,
    #[serde(default)]
    owner: String,
}

/// Read ~/.claude/tasks/{team}/*.json, compute pending/in_progress/completed and unblocked tasks.
fn get_available_tasks(team_name: &str) -> Option<TeamInfo> {
    let home = dirs::home_dir()?;
    let task_dir = home.join(".claude").join("tasks").join(team_name);
    let entries = std::fs::read_dir(&task_dir).ok()?;

    let mut tasks: Vec<TaskFile> = Vec::new();
    for entry in entries.flatten() {
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("json") {
            continue;
        }
        let data = match std::fs::read_to_string(&path) {
            Ok(d) => d,
            Err(_) => continue,
        };
        let t: TaskFile = match serde_json::from_str(&data) {
            Ok(t) => t,
            Err(_) => continue,
        };
        if !t.status.is_empty() {
            tasks.push(t);
        }
    }

    // Build completed ID set for dependency resolution
    let completed_ids: std::collections::HashSet<&str> = tasks
        .iter()
        .filter(|t| t.status == "completed")
        .map(|t| t.id.as_str())
        .collect();

    let mut info = TeamInfo {
        pending: 0,
        in_progress: 0,
        completed: 0,
        total: 0,
        unblocked: Vec::new(),
    };
    for t in &tasks {
        match t.status.as_str() {
            "completed" => info.completed += 1,
            "in_progress" => info.in_progress += 1,
            "pending" => {
                info.pending += 1;
                // Unblocked = all deps complete AND no owner assigned
                let all_deps_done = t
                    .blocked_by
                    .iter()
                    .all(|dep| completed_ids.contains(dep.as_str()));
                if all_deps_done && t.owner.is_empty() {
                    info.unblocked.push((t.id.clone(), t.subject.clone()));
                }
            }
            _ => continue,
        }
    }
    info.total = info.pending + info.in_progress + info.completed;
    Some(info)
}
